keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,510 @@
+      properties:
+        name:
+          type: string
+          default: default-name
+        number:
+          type: integer
+          default: 0
+```
+
+which would have the following behavior, both for CRD types and built-in types:
+```python
+>>> default('null')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: null}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {}}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {name: other-name}}')
+{"entry": {"name": "other-name", "number": 0}}
+```
+
+#### Struct pointers
+
+Struct pointers are allowed to have a specific default since they are distinct
+from unspecified non-pointers structs. Modifying the type from the
+previous example:
+```golang
+type Root struct {
+  // This is now a pointer, we can provide a default:
+  // +default={name: pointer-name}
+  Entry *SubLevel
+}
+
+type SubLevel struct {
+  // +defaut=default-name
+  Name string
+  // +default=0
+  Number int
+}
+```
+
+We would generate the following OpenAPI definition, both for CRD types (by kubebuilder)
+and built-in types:
+```yaml
+Root:
+  type: object
+  default: {}
+  properties:
+    entry:
+      $ref: SubLevel
+      default: {name: pointer-name}
+SubLevel:
+  type: object
+  properties:
+    name:
+      type: string
+      default: default-name
+    number:
+      type: integer
+      default: 0
+```
+
+which would have the following behavior, both for CRD types and built-in types:
+```python
+>>> default('null')
+{"entry": {"name": "pointer-name", "number": 0}}
+>>> default('{}')
+{"entry": {"name": "pointer-name", "number": 0}}
+>>> default('{entry: null}')
+{"entry": {"name": "pointer-name", "number": 0}}
+>>> default('{entry: {}}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {name: other-name}}')
+{"entry": {"name": "other-name", "number": 0}}
+```
+
+### Marker format
+
+The marker will have the following format:
+```
+// +default=<any>
+```
+which describes the value that the server will use if the field has a
+zero-value.
+
+`<any>` has to be in one-line YAML format (including its json form),
+representing the actual value to use by default. That is different from
+the current encoding in kubebuilder, but there seems to be consensus
+that they would be able to change that to YAML too for that new marker.
+
+Namely, this will allow to preserve the concise semantics that you want
+for most values:
+
+```golang
+type Foo struct {
+  // +default=32
+  Integer int
+
+  // +default=bar
+  String string
+
+  // +default=[popcorn, chips]
+  StringList []string
+}
+```
+while still allowing more complex expressions:
+```
+// +default=[{port: 80, name: http}]
+```
+
+Forcing the "one-line" subset allows us to sidestep a lot of the YAML
+weirdness in syntax, but we can still just use a YAML parser to parse
+the expressions if we want.
```

Comment:
a detail: null is allowed in default values?

Also already mentioned above somewhere: we should only allow validating values. We will see how clever the generator and kubebuilder can be made to check that (best effort, doesn't have to be perfect). CRDs will fails to be created if the value does not validate.